services/consensusalgo/benchmarkconsensus/leader.go

```diff
@@ -0,0 +1,198 @@
+	"github.com/orbs-network/orbs-network-go/crypto/signature"
+	"github.com/orbs-network/orbs-spec/types/go/protocol"
+	"github.com/orbs-network/orbs-spec/types/go/protocol/consensus"
+	"github.com/orbs-network/orbs-spec/types/go/protocol/gossipmessages"
+	"github.com/orbs-network/orbs-spec/types/go/services"
+	"github.com/orbs-network/orbs-spec/types/go/services/gossiptopics"
+	"github.com/pkg/errors"
+	"math"
+	"time"
+)
+
+func (s *service) leaderConsensusRoundRunLoop(ctx context.Context) {
+	s.lastCommittedBlock = s.leaderGenerateGenesisBlock()
+	for {
+		err := s.leaderConsensusRoundTick()
+		if err != nil {
+			s.reporting.Error(err)
+		}
+		select {
+		case <-ctx.Done():
+			s.reporting.Info("Consensus round run loop terminating with context")
+			return
+		case s.lastSuccessfullyVotedBlock = <-s.successfullyVotedBlocks:
+			s.reporting.Infof("Consensus round waking up after successfully voted block %d", s.lastSuccessfullyVotedBlock)
+			continue
+		case <-time.After(time.Duration(s.config.BenchmarkConsensusRoundRetryIntervalMillisec()) * time.Millisecond):
+			s.reporting.Info("Consensus round waking up after retry timeout")
+			continue
+		}
+	}
+}
+
+func (s *service) leaderConsensusRoundTick() (err error) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	// check if we need to move to next block
+	if s.lastSuccessfullyVotedBlock == s.lastCommittedBlockHeight() {
+		proposedBlock, err := s.leaderGenerateNewProposedBlockUnsafe()
+		if err != nil {
+			return err
+		}
+		err = s.saveToBlockStorage(proposedBlock)
+		if err != nil {
+			return err
+		}
+
+		s.lastCommittedBlock = proposedBlock
+		s.lastCommittedBlockVoters = make(map[string]bool)
+	}
+
+	// broadcast the commit via gossip for last committed block
+	_, err = s.gossip.BroadcastBenchmarkConsensusCommit(&gossiptopics.BenchmarkConsensusCommitInput{
+		Message: &gossipmessages.BenchmarkConsensusCommitMessage{
+			BlockPair: s.lastCommittedBlock,
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// used for the first commit a leader does which is nop (genesis block) just to see where everybody's at
+func (s *service) leaderGenerateGenesisBlock() *protocol.BlockPairContainer {
+	transactionsBlock := &protocol.TransactionsBlockContainer{
+		Header:             (&protocol.TransactionsBlockHeaderBuilder{BlockHeight: 0}).Build(),
+		Metadata:           (&protocol.TransactionsBlockMetadataBuilder{}).Build(),
+		SignedTransactions: nil,
+		BlockProof:         nil,
+	}
+	resultsBlock := &protocol.ResultsBlockContainer{
+		Header:              (&protocol.ResultsBlockHeaderBuilder{BlockHeight: 0}).Build(),
+		TransactionReceipts: nil,
+		ContractStateDiffs:  nil,
+		BlockProof:          nil,
+	}
+	blockPair, err := s.leaderSignBlockProposal(transactionsBlock, resultsBlock)
+	if err != nil {
+		panic(err)
+	}
+	return blockPair
+}
+
+func (s *service) leaderGenerateNewProposedBlockUnsafe() (*protocol.BlockPairContainer, error) {
```

Comment:
why 'unsafe'?

Reply:
Unsafe means that it doesn't capture mutex and assumes the mutex is captured. Meaning if you don't have the mutex, don't call an unsafe function

Reply:
Rename to "UnderMutex"